Tasks/CargoAuthenticateV0/cargoauthenticatemain.ts

```diff
@@ -79,26 +87,49 @@ async function main(): Promise<void> {
             const registryConfigName = registry.toLocaleUpperCase().replace(/-/g, "_");
             const tokenName = `CARGO_REGISTRIES_${registryConfigName}_TOKEN`;
             const credProviderName = `CARGO_REGISTRIES_${registryConfigName}_CREDENTIAL_PROVIDER`;
+            let connectionType = '';
             if (tl.getVariable(tokenName)) {
-                tl.debug(tl.loc('ConnectionAlreadySet', registry, tl.getVariable(tokenName).indexOf('Basic') !== -1 ? 'external' : 'internal'));
+                connectionType = tl.getVariable(tokenName).indexOf('Basic') !== -1 ? 'external or federated' : 'internal';
+                tl.debug(tl.loc('ConnectionAlreadySet', registry, connectionType));
             }
-            if (registryUrl && registryUrl.host && collectionHosts.indexOf(registryUrl.host.toLowerCase()) >= 0) {
+            if (registryUrl && registryUrl.host && collectionHosts.indexOf(registryUrl.host.toLowerCase()) >= 0 && (connectionType !== 'external or federated')) {
                 let currentRegistry : string;
+                const currentRegistryUrl = url.parse(result.registries[registry].index.replace("sparse+", "")).href;
+#if WIF
```

Comment:
Since Cargo needs to parse config.toml for this to work, if a user has two feed in config.toml that they want to authenticate with federated token, there will be two consecutive runs of CargoAuthenticate. How is the current implementation making sure the second run doesn't overwrite the first run's feed URL federated token with system access token?

Reply:
The last part of the if statement on line 95 checks if the registry was already authenticated with an external or federated token, and if it was, it skips over that registry

Reply:
We may need to rethink how we are doing this in Cargo. The current implementation is also missing the check where the feed URl doesn't match any registry in the config.toml.

Reply:
And what if user passes in a feed URL, but it's not part of their config.toml? We dont have a registry name in this case

Reply:
I see what you're saying. In my last changes I reorganized this whole section so that it comes earlier. I also added a check to make sure the feedUrl is in the config.toml and then it only authenticates the corresponding registry.

Reply:
Switched this approach from using feedUrls to using registry names and ensuring they exist in the config.toml